rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -2987,6 +3020,87 @@ extern "C" rmw_ret_t rmw_get_node_names(
   return RMW_RET_BAD_ALLOC;
 }
 
+extern "C" rmw_ret_t rmw_get_node_names_with_security_contexts(
+  const rmw_node_t * node,
+  rcutils_string_array_t * node_names,
+  rcutils_string_array_t * node_namespaces,
+  rcutils_string_array_t * security_contexts)
+{
+  RET_WRONG_IMPLID(node);
+  auto node_impl = static_cast<CddsNode *>(node->data);
+  if (rmw_check_zero_rmw_string_array(node_names) != RMW_RET_OK ||
+    rmw_check_zero_rmw_string_array(node_namespaces) != RMW_RET_OK)
+  {
+    return RMW_RET_ERROR;
+  }
+
+  std::vector<std::tuple<std::string, std::string, std::string>> ns;
+  const auto re = std::regex(
+    "^name=([^;]*);namespace=([^;]*);securitycontext=([^;]*);", std::regex::extended);
+  auto oper =
+    [&ns, re](const dds_builtintopic_participant_t & sample, const char * ud) -> bool {
+      std::cmatch cm;
+      static_cast<void>(sample);
+      if (std::regex_search(ud, cm, re)) {
+        ns.push_back(std::make_tuple(std::string(cm[1]), std::string(cm[2]), std::string(cm[3])));
+      }
+      return true;
+    };
+  rmw_ret_t ret;
+  if ((ret = do_for_node_user_data(node_impl, oper)) != RMW_RET_OK) {
+    return ret;
+  }
+
+  rcutils_allocator_t allocator = rcutils_get_default_allocator();
+  if (rcutils_string_array_init(node_names, ns.size(), &allocator) != RCUTILS_RET_OK ||
+    rcutils_string_array_init(node_namespaces, ns.size(), &allocator) != RCUTILS_RET_OK ||
+    rcutils_string_array_init(security_contexts, ns.size(), &allocator) != RCUTILS_RET_OK)
+  {
+    RMW_SET_ERROR_MSG(rcutils_get_error_string().str);
+    goto fail_alloc;
+  }
+  size_t i;
+  i = 0;
```

Comment:
@ivanpauno nit: last two lines can be collapsed into one.

Reply:
Same as above, it comes from `rmw_get_node_names`.

Reply:
There is a pretty fun reason: when done in the same line, the compiler fails saying that the above `goto` is skipping an initialization. In general, it's safe to skip the initialization of a trivially destructible object (AFAIU), but the compiler shows the warning anyways (and we transform that warning in an error, because of our compiler flags).

When split in two lines, the compiler doesn't show a warning.

P.S.: This kind of warning never happens in pure C code, it only applies when you use C++.

Reply:
Hmm interesting, odd but interesting.

Reply:
Please define ‘i’ higher in the scope to avoid this instead. This what is done in all the other functions that use goto.

Reply:
I think @wjwwood still applies.

Reply:
The code is pre-existent, and I don't want to retrigger CI because of this. If you don't mind, I will ignore the comment.